### PR TITLE
HSM: add shared interface and factory pattern for SHA algorithms

### PR DESCRIPTION
**Pull request by mali053**

### CODE REVIEW

File: include/sha256.h [hunk 1]
```diff
@@ -1,23 +1,43 @@
...
-    size_t bitLength = 0;      // Total length of the message in bits
-};
-
-// Function prototypes for SHA-256 hashing
-void sha256_update(SHA256State& state, const std::vector<uint8_t>& data);
-void padding(SHA256State& state);
-std::vector<uint8_t> sha256_finalize(SHA256State& state);
-std::vector<uint8_t> sha256_compute(const std::vector<uint8_t>& input);
-void transform(SHA256State& state);
+// #include "return_codes.h"
```
Comment (TobyRoth) [inline include/sha256.h]: Why is it commented?

File: tests/ecc_tests.cpp [hunk 2]
```diff
@@ -19,8 +19,8 @@ TEST(ECCTest, EncryptDecrypt)
...
-{
-    ::testing::InitGoogleTest(&argc, argv);
-    return RUN_ALL_TESTS();
-}
\ No newline at end of file
+// int main(int argc, char **argv)
+// {
+//     ::testing::InitGoogleTest(&argc, argv);
+//     return RUN_ALL_TESTS();
+// }
```
Comment (TobyRoth) [inline tests/ecc_tests.cpp]: What's the point of this change and why is it related to this commit?

File: tests/aes_tests.cpp [hunk 3]
```diff
@@ -118,12 +118,3 @@ TEST(KeyLengths, KeyLength256_CTR)
...
 }
-
-int main() {
-    // Register factories
-    FactoryManager& manager = FactoryManager::getInstance();
-
-    // Now run your tests
-    ::testing::InitGoogleTest();
-    return RUN_ALL_TESTS();
-}
```
Comment (TobyRoth) [inline tests/aes_tests.cpp:129]: What's the point of this change and why is it related to this commit?

File: tests/hash_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
+#include <gtest/gtest.h>
+#include <memory>
+#include <vector>
+#include "../include/IHash.h"       // בהתאם למיקום שלך
+#include "../include/hash_factory.h" // בהתאם למיקום שלך
```
Comment (TobyRoth) [inline tests/hash_tests.cpp]: Please remove...

File: tests/hash_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    std::string hash = hashStringSHA256("abc");
+    EXPECT_EQ(hash, expectedHash);
+}
+// Test hashing "hello world"
+TEST_F(HashTest, HashHelloWorld) {
+    std::string expectedHash = "b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9";
+    std::string hash = hashStringSHA256("hello world");
+    EXPECT_EQ(hash, expectedHash);
+}
+// Test hashing a string with special characters
```
Comment (TobyRoth) [inline tests/hash_tests.cpp:79]: Add a blank line before

File: tests/hash_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    std::string hash = hashStringSHA256("");
+    EXPECT_EQ(hash, expectedHash);
+}
+// Test hashing "abc"
+TEST_F(HashTest, HashABC) {
+    std::string expectedHash = "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad";
+    std::string hash = hashStringSHA256("abc");
+    EXPECT_EQ(hash, expectedHash);
+}
+// Test hashing "hello world"
```
Comment (TobyRoth) [inline tests/hash_tests.cpp:72]: Add a blank line before

File: tests/hash_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    std::unique_ptr<IHash> sha256;
+    std::unique_ptr<IHash> sha512;
+};
+// Test hashing an empty string
+TEST_F(HashTest, HashEmptyString) {
+    std::string expectedHash = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
+    std::string hash = hashStringSHA256("");
+    EXPECT_EQ(hash, expectedHash);
+}
+// Test hashing "abc"
```
Comment (TobyRoth) [inline tests/hash_tests.cpp:65]: Add a blank line before

File: tests/hash_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        sha512 = std::unique_ptr<IHash>(factoryManager->create(IHash::SHA3_512));
+        // sha256.reset(factoryManager->create(IHash::SHA256));
+        if (!sha256) {
+            FAIL() << "Failed to create SHA256 instance";
+        }
+        if(!sha512){
+            FAIL() << "Failed to create SHA512 instance";
+        }
+    }
+    // Utility to hash a string and return the hex string of the result
```
Comment (TobyRoth) [inline tests/hash_tests.cpp:34]: Add a blank line before

File: tests/hash_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        sha512->finalize(output);
+        std::string res = std::string(output.begin(), output.end());
+
+        ASSERT_EQ(res, expectedHash);
+    }
+    HashFactory* factoryManager;
+    std::unique_ptr<IHash> sha256;
+    std::unique_ptr<IHash> sha512;
+};
+// Test hashing an empty string
```
Comment (TobyRoth) [inline tests/hash_tests.cpp:58]: Add a blank line before

**Comment by mali053** [comment]
I uploaded the changes, want to go over it?

**Pull request closed without merging**